apps/meteor/app/e2e/server/functions/fetchUsersWaitingForGroupKey.ts

```diff
@@ -0,0 +1,46 @@
+import type { IRoom, IUser } from '@rocket.chat/core-typings';
+import { Rooms, Subscriptions, Users } from '@rocket.chat/models';
+
+import { isTruthy } from '../../../../lib/isTruthy';
+
+const NUMBER_OF_ROOMS = 10;
+const QUEUE_SIZE = 50;
+
+export async function fetchUsersWaitingForGroupKey(userId: IUser['_id']): Promise<{
+	usersWaitingForE2EKeys: Record<IRoom['_id'], { _id: IUser['_id']; public_key: string }[]>;
+	hasMore: boolean;
+}> {
+	const roomIds = (await Subscriptions.findByUserId(userId, { projection: { rid: 1 } }).toArray()).map((sub) => sub.rid);
+
+	const cursor = Rooms.findRoomsByRoomIdsWithE2EEQueue(roomIds, { projection: { usersWaitingForE2EKeys: 1 } });
+
+	const count = await cursor.count();
+
+	const hasMore = count > NUMBER_OF_ROOMS;
+	const encryptedRoomIds = (await cursor.toArray()).map((room) => room._id);
+
+	const rooms = await Rooms.fetchRandomUsersFromE2EEQueue(encryptedRoomIds, NUMBER_OF_ROOMS, QUEUE_SIZE);
```

Comment:
There can be many rooms, we are just getting random 10 rooms, not sure if mongo returns rooms randomly 🤔

Reply:
@KevLehman because many users can be doing this operation at the same time so it try to prevent them to generate keys for the same users and collaborate to generate to more users at the same time if it happens

Reply:
Thanks for helping my curiosity.